Add style to pagination component

CharlyMartin opened the pull request:
This PR adds styling, ellipsis, and accessibility concerns to the pagination component.

mirhamasala commented:
Super cool @CharlyMartin! ☀️ As early review comments go:

- Let's make this keyboard accessible
- Let's work on matching the styles closer to the Figma design (Vertical lines are a different color, chevrons seems a little thicker, and there's not so much whitespace around the numbers.)

<img width="505" alt="Screenshot 2024-04-12 at 07 23 37" src="https://github.com/FilecoinFoundationWeb/filecoin-foundation/assets/26620750/492f9d6c-726b-4760-b4da-0f1048ae27f1">

<img width="1183" alt="Screenshot 2024-04-12 at 07 22 56" src="https://github.com/FilecoinFoundationWeb/filecoin-foundation/assets/26620750/8be63080-ae0e-4213-b37e-feacdbaf3630">

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -13,7 +17,13 @@ export function Pagination({
+  const collapsedRange = useCollapseRange(
+    pageCount,
+    currentPage,
+    isXsDevice ? 2 : 6,
```
The `2` and `6` are like[ magic numbers](https://www.inf.unibz.it/~calvanese/teaching/05-06-ip/lecture-notes/uni04/node17.html#:~:text=A%20magic%20number%20is%20a,difficult%20to%20maintain%20and%20update.) - We can define them with variables.

CharlyMartin replied:
Love it! 

I added a `useResponsiveRange` to compute the range's size based on the screen size. Hopefully, the numbers make more sense in that context.

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -28,37 +38,63 @@ export function Pagination({
-      </button>
+      <ul className="flex shrink grow justify-center gap-2 sm:gap-1">
+        {collapsedRange.map((item, index) => {
+          return (
```
We can use an implicit return here.

CharlyMartin replied:
✅

mirhamasala commented on src/app/_components/ScreenSizeIndicator.tsx:
```diff
@@ -0,0 +1,14 @@
+export function ScreenSizeIndicator() {
+  if (process.env.NODE_ENV === 'production') return null
+
+  return (
+    <div className="fixed bottom-1 left-1 z-50 rounded bg-zinc-800 px-1 font-mono text-xs text-white dark:bg-zinc-100 dark:text-zinc-900">
+      <div className="block sm:hidden">xs</div>
+      <div className="hidden sm:block md:hidden">sm</div>
+      <div className="hidden md:block lg:hidden">md</div>
+      <div className="hidden lg:block xl:hidden">lg</div>
+      <div className="hidden xl:block 2xl:hidden">xl</div>
+      <div className="hidden 2xl:block">2xl</div>
+    </div>
+  )
+}
```
Should be a different PR.

CharlyMartin replied:
Yes, you're right. I reverted it.

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+    )
+  }
+
+  // Ellipsis at the start only
+  if (needsStartEllipsis && !needsEndEllipsis) {
```
If we're writing comments, it's a good indication that the code could be clearer.

This condition, `needsStartEllipsis && !needsEndEllipsis`, can be defined in a variable.

CharlyMartin replied:
> If we're writing comments, it's a good indication that the code could be clearer.

I see what you mean, but for long/complex code, I think comments add a visual hierarchy to the file so that it's more accessible for the reader, in the same way headings structure paragraphs.

Thoughs?

mirhamasala replied:
I don't agree. If the code is long and complex, then it should be broken up in shorter and simpler code. Abstract bits into functions, etc. I recommend reading chapter 4 from Clean Code by Robert C. Martin. (I recommend the entire book - until about chapter 8 - After that, it's very Java specific if I remember correctly.)

>So when you find yourself in a position where you need to write a comment, think it through and see whether there isn’t some way to turn the tables and express yourself in code. Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+import { useMemo } from 'react'
+
+// Inspiration: https://atlassian.design/components/pagination
```
```suggestion
```

mirhamasala replied:
Can be added to the PR message if needed.

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+    return Array.from({ length: pageCount }, (_, i) => i + 1)
+  }, [pageCount])
+
+  // Boolean
```
```suggestion
```

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+  const needsStartEllipsis = currentPage - DISTANCE_FROM_START > 0
+  const needsEndEllipsis = currentPage + DISTANCE_FROM_END < pageNumbers.length
+
+  // page numbers
```
```suggestion
```

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+  const firstPageNumber = pageNumbers[0]
+  const lastPageNumber = pageNumbers[pageNumbers.length - 1]
+
+  // Indexes of page numbers
```
```suggestion
```

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+  if (!needEllipsis) return pageNumbers
+  if (range === 1) return [currentPage]
+
+  // If range is less than 4, we don't show any ellipsis
```
Instead of having comments, could we abstract this into a function with appropriate naming?

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+
+  // Ellipsis at the start only
+  if (needsStartEllipsis && !needsEndEllipsis) {
+    const remainingArraySlots = range - 2
```
Magic numbers floating around `2` and `4`  ✨

CharlyMartin replied:
What do you mean?

mirhamasala commented on src/app/_hooks/useCollapseRange.ts:
```diff
@@ -0,0 +1,85 @@
+export function useCollapseRange(
+  pageCount: number,
+  currentPage: number,
+  range: number = 7,
```
Magic number ✨ `7`

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -28,37 +38,63 @@ export function Pagination({
+      <div className="flex">
+        <button
+          className="flex items-center rounded bg-brand-300 p-1 px-2 transition hover:bg-brand-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-white"
+          aria-label="Previous"
```
The `aria-label` could be more descriptive `Go to previous page`

CharlyMartin commented:
Thank you for the review @mirhamasala ✨ 

I've implemented most of your feedback with these 3 commits:
1. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/317c8538d7838d98f8144ed793b85caeb1188b4c
2. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/829ad66d606e4b0041e0d0ebf5f2bfb65ce8bdae
3. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/aea1d6a0dffad1ce0857a8dddc249481b3e79af8

I also removed the `ScreenSizeIndicator` with commit https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/bb8ba801fd73b1e9fa073bf6ddcd4d14ba8a5f45, we'll add it later via a separate [ticket](https://www.notion.so/filecoin/FF-V4-Add-ScreenSizeIndicator-component-e85bd71442af4670a40d52ac4cae86b5?pvs=4).

Regarding the keyboard accessibility of the page, something is off indeed so we can explore a fix in a separate [ticket](https://www.notion.so/filecoin/FF-V4-Explore-keyboard-accessibility-of-blog-page-and-pagination-component-981a71bbad4d4a8d91c8ee7868dd754b?pvs=4) as well

Finally, I added a few more comments/questions.

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -13,52 +17,110 @@ export function Pagination({
+          <LineVertical
+            size={24}
+            weight="light"
+            className="text-[#08072E]/50"
```
```suggestion
            className="text-brand-800/50"
```

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -13,52 +17,110 @@ export function Pagination({
+                    ? 'bg-brand-800 text-brand-100'
+                    : 'bg-brand-300 text-brand-700 hover:bg-brand-400',
+                )}
+                aria-label={'Page ' + item}
```
```suggestion
                aria-label={`Go to page ${item}`}
```

mirhamasala commented on src/app/_components/Pagination.tsx:
```diff
@@ -13,52 +17,110 @@ export function Pagination({
+          <LineVertical
+            size={24}
+            weight="light"
+            className="text-[#08072E]/50"
```
```suggestion
            className="text-brand-800/50"
```

CharlyMartin commented:
@mirhamasala I've reworked the `useCollapseRange` hook, now called `useVisiblePages`, to be more accessible and easier to read.

Let me know what you think :)

mirhamasala commented:
> Let me know what you think :)

@charlymartin - I think it's amazing. Thank you for making it happen. ✨
